File: functions/src/api/guild.ts
import { NotAuthorizedError, PermissionDeniedError, InvalidOperationError } from '../error';
import { onHttpsCall } from '../common';
import { User } from '../controller/user';
import { Guild } from '../controller/guild';

export const guild_getGuild = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const guild = Guild.get(data.gid);
    await guild.ensureDataExistInLocal();

    const users = await guild.getUsers();
    return {
        ...guild,
        users
    };
});
export const guild_queryGuild = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const guilds = await Guild.query(data.name);
    await Promise.all(
        guilds.map(x => x.ensureDataExistInLocal())
    );

    return {
        guilds
    };
});

export const guild_requestJoin = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const guild = Guild.get(data.gid);
    const user = User.get(ctx.auth.uid);

    await guild.requestJoin(user);

    return {
    };
});
export const guild_acceptRequest = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const guild = Guild.get(data.gid);
    if (guild.data!.owner != ctx.auth.uid)
        throw new PermissionDeniedError();

    const user = User.get(data.uid);
    await guild.acceptRequest(user);
    
    return {
    };
});
export const guild_rejectRequest = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const guild = Guild.get(data.gid);
    if (guild.data!.owner != ctx.auth.uid)
        throw new PermissionDeniedError();

    const user = User.get(data.uid);
    await guild.rejectRequest(user);
    
    return {
    };
});

export const guild_updateProperty = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const guild = Guild.get(data.gid);
    await guild.ensureDataExistInLocal();

    if (guild.data!.owner != ctx.auth.uid)
        throw new PermissionDeniedError();

    await guild.update(data);

    return {
        guild
    };
});
export const guild_transferOwnership = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const guild = Guild.get(data.gid);
    await guild.ensureDataExistInLocal();

    if (guild.data!.owner != ctx.auth.uid)
        throw new PermissionDeniedError();

    await guild.update({
        owner: data.uid
    });

    return {
        guild
    };
});
export const guild_grantSubownership = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const guild = Guild.get(data.id);
    await guild.ensureDataExistInLocal();

    if (guild.data!.owner != ctx.auth.uid)
        throw new PermissionDeniedError();

    if (!guild.data!.subOwner1) {
        await guild.update({
            subOwner1: data.uid
        });    
    }
    else if (!guild.data!.subOwner2) {
        await guild.update({
            subOwner2: data.uid
        });    
    }
    else 
        throw new InvalidOperationError('No place to add');

    return {
        guild
    };
});

File: sdk/FriendKit/Assets/FriendKit/Guild.cs
using System.Collections;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Threading.Tasks;
using UnityEngine;

using Newtonsoft.Json;

namespace FriendKit
{
    public class GuildModel
    {
        public string id;
        public string name;

        public UserModel[] users;
    }
    public class Guild : GuildModel
    {
        public async static Task<Guild> Get(string id)
        {
            var func = FB.Functions;
            var resp = await func.GetHttpsCallable("guild_getGuild")
                .CallAsync(new Dictionary<string, object>()
                {
                    ["gid"] = id
                });

            return resp.Data.Reinterpret<GetGuildResponse>().guild;
        }
        public async static Task<Guild[]> Query(string name)
        {
            var func = FB.Functions;
            var resp = await func.GetHttpsCallable("guild_queryGuild")
                .CallAsync(new Dictionary<string, object>()
                {
                    ["name"] = name
                });

            return resp.Data.Reinterpret<GetGuildsResponse>().guilds;
        }

        public async Task Update(GuildModel guild)
        {
            var func = FB.Functions;
            var resp = await func.GetHttpsCallable("guild_updateProperty")
                .CallAsync(new Dictionary<string, object>()
                {
                    ["name"] = guild.name
                });
        }
        public async Task TransferOwnership(User user)
        {
            var func = FB.Functions;
            var resp = await func.GetHttpsCallable("guild_transferOwnership")
                .CallAsync(new Dictionary<string, object>()
                {
                    ["gid"] = id,
                    ["uid"] = user.id
                });
        }
        public async Task GrantSubownership(User user)
        {
            var func = FB.Functions;
            var resp = await func.GetHttpsCallable("guild_grantSubownership")
                .CallAsync(new Dictionary<string, object>()
                {
                    ["gid"] = id,
                    ["uid"] = user.id
                });
        }

        public async Task AcceptRequest(User user)
        {
            var func = FB.Functions;
            var resp = await func.GetHttpsCallable("guild_acceptRequest")
                .CallAsync(new Dictionary<string, object>()
                {
                    ["gid"] = id,
                    ["uid"] = user.id
                });
        }
        public async Task RejectRequest(User user)
        {
            var func = FB.Functions;
            var resp = await func.GetHttpsCallable("guild_rejectRequest")
                .CallAsync(new Dictionary<string, object>()
                {
                    ["gid"] = id,
                    ["uid"] = user.id
                });
        }
    }

    class GetGuildResponse
    {
        public Guild guild;
    }
    class GetGuildsResponse
    {
        public Guild[] guilds;
    }
}